server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java

```diff
@@ -916,6 +919,47 @@ public ReplicationGroup getReplicationGroup() {
         return replicationGroup;
     }
 
+    /**
+     * Returns the pending replication actions for the shard.
+     *
+     * @return the pending replication actions
+     */
+    public PendingReplicationActions getPendingReplicationActions() {
+        assert primaryMode;
+        return pendingReplication;
+    }
+
+    private void updateReplicationGroupAndNotify(boolean replicasMightHaveChanged) {
+        ReplicationGroup newReplicationGroup = calculateReplicationGroup();
+        ReplicationGroup oldReplicationGroup = this.replicationGroup;
+        if (replicasMightHaveChanged) {
+            Set<String> oldReplicaNodeIds;
+            if (oldReplicationGroup == null) {
+                oldReplicaNodeIds = Collections.emptySet();
+            } else {
+                oldReplicaNodeIds = oldReplicationGroup.getReplicationTargets().stream()
+                    .map(ShardRouting::currentNodeId)
+                    .collect(Collectors.toSet());
+            }
+            for (ShardRouting replica : newReplicationGroup.getReplicationTargets()) {
+                if (oldReplicaNodeIds.contains(replica.currentNodeId()) == false) {
+                    pendingReplication.nodeJoinedReplicationGroup(replica.currentNodeId());
+                }
+            }
+        }
+        this.replicationGroup = newReplicationGroup;
```

Comment:
I think that I would prefer for this to be set before we call the methods on `pendingReplication`, just to be sure that an exception there does not mess up the state in this class.

Reply:
This work was to prevent the race where the `ReplicationGroup` is returned by the replication tracker with a new node to a `ReplicationOperation`. This operations attempts to start a replication request, but the listener to `PendingReplicationActions` has not yet been called so the request is immediately cancelled.

I made your change. But I think spurious cancellations are possible.

Reply:
ok, I see now. One option is to turn it around then, but use `try - finally` to make sure that the state is updated even when the listener throws.

Another option (currently preferring this one for its generality) is that we have a versioning concept on `ReplicationGroup` (i.e. knowing which one is newer than another one) and that we explicitly update `PendingReplicationActions` whenever we capture the `ReplicationGroup` in `IndexShard` (most times the update will be a NOOP and should not need any locking on `PendingReplicationActions`).